### DISCUSSION TITLE
Authorization challenge for genuine products

### DISCUSSION THREAD

**Discussion by leotordo** [HomeSpan/HomeSpan#559] [Q&A]
HI,
  a client of mine would like to buy my project made with homespan. However, he would like the board to be recognized as compatible by the Home app.
I'm gathering information. In practice, Apple sells an IC based on I2C bus that manages the authenticity of the product.
In practice, the "Home app" sends a data packet that must be sent to the coprocessor. This responds with a data packet that needs to be transmitted back to the "Home app".
In practice, the ESP32 must act as a bridge between the apple protocol and the I2C bus.
Being on I2C I don't think there are hardware problems: ESP32 manages this protocol and also homespan (if I read the FAQ correctly).
Can anyone tell me if there is any way to figure out when the app sends the authentication request (Apple calls it an authorization challenge) and how to handle it?

Thank you all
Stefano

**Comment by HomeSpan** [comment]
This can only be done with an MFi license from Apple.  Prior to 2019 it required an Apple co-processor that would communicate via I2C.  After 2019 manufacturers could authenticate in software only (without the need for the co-processor).  

All HomeKit devices advertise their services via MDNS.  The "ff" text field is used to specify the type of device.  For non-commercial devices using HAP-R2, the "ff" field is broadcast as zero.  This indicates a non-commercial device to be paired without MFi authentication.  When the Home App tries to pair with such a device it will follow the protocol specified in HAP-R2 (which HomeSpan follows).  If you broadcast something other than zero in the "ff" field, the Home App will try to pair with the HomeKit device using either hardware or software authentication.  There are no public specs on how this works (nor even a list of the possible "ff" fields).

**Reply by leotordo** [reply]
Yes, I also understand how authentication via MFi license works.
My client has the apple account for application development and sent me some specifications (Accessory Interface
specification). Hardware coprocessor is required

Here an excerpt
Accessory authentication provides a mechanism for devices to trust the identities and supported
features of specification compliant accessories.
Accessory authentication requires an Apple Authentication Coprocessor, a hardware component
containing an X.509 certificate and the ability to generate responses to authentication challenges.
An authentication challenge is a random value sent from a device to an accessory. The accessories shall
perform a computation on the offered challenge and return a result, that is, the challenge response, to
the device for verification.
This feature is supported on tvOS 10.2 or later

The hardware coprocessor can only be bought if you have an apple account, I inserted it into the hardware connected to the ESP32 I2C.
I need to figure out how to receive the request from the app and how to send the app the packet processed by the hardware coprocessor.

Any ideas?

**Reply by josephcherayil** [reply]
Hi 
I'm also interested in MFI certification.can you please post the specs photos of that hardware supplied by Apple.

**Reply by leotordo** [reply]
Hello Joseph,
I finally solved the auth. challenge problem.
The way is not to use Homespan :-(
Apple said me that the library used for Homespan is not compliant for commercial purpose.

I made affiliation to MFi program and get my ID. Next I got  [Espressif Homekit SDK ](https://www.espressif.com/en/products/sdks/esp-homekit-sdk) by sending a request (with MFi informations)

Now I'm developping the application with it
Regards

**Reply by itatboy** [reply]
Hi, leotordo,
I am developing Homekit accessories and need to add MFI certification. Can you provide a general description of the steps to join the MFI program and any special requirements? Thank you.

**Reply by leotordo** [reply]
@itatboy 
HomeSpan is not compatible with MFi certification. It's very good for non-commercial applications.

As I wrote you have to subscribe to MFi Program (https://mfi.apple.com/).
Then with your MFi account you can access for free to esp-homekit-sdk (for MFi). I use this sdk with VScode (you can find a lot of tutorial on youtube)

For your hardware you need the authentication coprocessor. You can buy it from distribution but you need the MFi code to buy it

Hope I've help you
bye
Leo

**Reply by itatboy** [reply]
Thank you very much for your reply, it has been very helpful to me.

As far as I know, MFI authentication can be verified through software and no longer requires an authentication coprocessor. Is the information I understand incorrect?

Can you provide more information about MFI? Because besides the website you provided, there is too little information about MFI.I hope to receive more information on how to join the MFI program.

Here is some known information: MFI has two types of licenses: developer license and manufacturing license. Can individual developers obtain a developer license to obtain the Espressif Homekit SDK? If the developer license is only valid for one year, will the SDK become invalid?

Thanks，Leo

**Reply by leotordo** [reply]
@itatboy I don't known anything about software authentication, sorry. I began with HW coprocessor and it worked, so I didn't search informations about SW auth.
I've no additional information about MFi program and I can't answer your questions. I think you've to write a info request to Apple
Regards

**Reply by itatboy** [reply]
Received a response, thank you

**Comment by leotordo** [comment]
The hardware is very simple. Two pins for power supply, two pins for I2C bus

![image](https://user-images.githubusercontent.com/121090098/234856305-6fd6b12a-33db-4aa3-9edb-e9e87755977f.png)

I'm waiting for our buyer dept for samples.

The question is the software handling of this device, within ESP32 and Homekit protocol used by Homespan

Do you have some information about it?
Thanks
Stefano

**Reply by leotordo** [reply]
Finally I've the prototype with Auth Coprocessor (CP). I tried some read/write request from a PIC microcontroller I've got on the same I2C but and the CP works and replies correctly.
Now the question: how to implement authentication challenge on HomeSpan?
